File: Tree/04. Maximum Depth of Binary Tree/04. Maximum Depth of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# 递归解法：树root的最大深度=max(左子树root.left的最大深度, 右子树root.right的最大深度)+1

class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        elif not root.left and not root.right:
            return 1
        else:
            return max(self.maxDepth(root.left),self.maxDepth(root.right))+1
        

# 非递归解法：使用层次遍历，在层次遍历的过程中记录队列中各个结点的深度信息，从而计算得到最大深度

class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        
        que=[]    # 用于层次遍历的队列
        que.append([root,1])   # 在队列中记录root的深度为1

        depth=0   # 记录最终的树的最大深度

        while que:
            temp=que[0]
            que.pop(0)

            depth=temp[1] if temp[1]>depth else depth

            if temp[0].left:
                que.append([temp[0].left, temp[1]+1])
            if temp[0].right:
                que.append([temp[0].right, temp[1]+1])
        
        return depth

File: Tree/04. Maximum Depth of Binary Tree/04. Maximum Depth of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// 二叉树中的基本题型：使用递归即可，一个二叉树root的最大深度等于左子树root.left的深度以及右子树root.right中的深度中的最大值加上1


// 递归解法：
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if(!root)
        return 0;
        else if(!root->left&&!root->right)
        return 1;
        else
        {
            int leftdepth=maxDepth(root->left);
            int rightdepth=maxDepth(root->right);

            return ((leftdepth>rightdepth)?leftdepth:rightdepth)+1;
        }
    }
};


// 下面同样提供一种非递归解法，面试中，树的递归方法的非递归实现也是一个考点
// 非递归算法思想：可以利用树的层次遍历方法，使用一个队列实现层次遍历时，向队列中加入树的结点时同样加入深度信息，从而进行求解

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if(!root)
        return 0;
        
        queue<pair<TreeNode*, int> > que;   //用于层次遍历的栈
        que.push(make_pair(root,1));        //将根节点加入队列中

        int depth=1;                        //记录树的最大深度

        while(que.size())
        {
            // 从队列中出队一个树的结点并输出，更新树的最大深度
            pair<TreeNode*, int> temp=que.front();
            que.pop();

            depth=(temp.second>depth)?temp.second:depth;

            // 将出队的结点的左子结点和右子结点加入队列，入队时，记录这两个结点的深度+1
            if(temp.first->left)    // 注意空结点无需加入队列
            que.push(make_pair(temp.first->left,temp.second+1));
            if(temp.first->right)   // 注意空结点无需加入队列
            que.push(make_pair(temp.first->right,temp.second+1));
        }

        return depth;
    }
};